alf/algorithms/rl_algorithm.py

```diff
@@ -563,20 +582,195 @@ def _train_iter_off_policy(self):
 
-        with torch.set_grad_enabled(config.unroll_with_grad):
-            with record_time("time/unroll"):
-                self.eval()
-                experience = self.unroll(config.unroll_length)
-                self.summarize_rollout(experience)
-                self.summarize_metrics()
+        if self._count == 0 or self._count > self._rl_train_after_update_steps:
+            # 1) RL training
+            with torch.set_grad_enabled(config.unroll_with_grad):
+                with record_time("time/unroll"):
+                    self.eval()
+                    experience = self.unroll(config.unroll_length)
+                    self.summarize_rollout(experience)
+                    self.summarize_metrics()
+        else:
+            experience = None
 
         self.train()
         steps = self.train_from_replay_buffer(update_global_counter=True)
 
         with record_time("time/after_train_iter"):
-            train_info = experience.rollout_info
-            experience = experience._replace(rollout_info=())
+            if experience is not None:
+                train_info = experience.rollout_info
+                experience = experience._replace(rollout_info=())
+            else:
+                experience = None
+                train_info = None
             self.after_train_iter(experience, train_info)
 
         # For now, we only return the steps of the primary algorithm's training
         return steps
+
+    def load_offline_replay_buffer(self, untransformed_observation_spec):
+        """Load replay buffer from a replay buffer checkpoint.
+        It will construct a replay buffer (``self._offline_replay_buffer``)
+        holding the data loaded from the checkpoint, which can be used for
+        model training, e.g. in the hybrid training pipeline or in other ways.
+
+        Args:
+            untransformed_observation_spec (nested TensorSpec): spec that
+                describes the strcuture of the utransformed observations.
+        """
+
+        if self._offline_buffer_dir is None or self._offline_buffer_dir == "":
+            # no offline buffer is provided
+            return
+        else:
+            logging.info('------offline replay buffer loading started------')
+
+            map_location = None
+            if not torch.cuda.is_available():
+                map_location = torch.device('cpu')
+            replay_buffer_checkpoint = torch.load(
+                self._offline_buffer_dir, map_location=map_location)
+
+            buffer_dict = replay_buffer_checkpoint['algorithm']
+
+            def _get_full_key(dict, partial_key):
+                full_key = next((key for key in dict if partial_key in key),
+                                None)
+                assert full_key is not None, (
+                    "key containing {} "
+                    "is not found.".format(partial_key))
+                return full_key
+
+            k1 = _get_full_key(buffer_dict, "time_step|reward")
```

Comment:
what is the use of k1?

Reply:
This line was a debugging line. Removed.